File: app/python/auth/modules/bearer_static.py
from __future__ import annotations

import os
from typing import Any

import httpx
from redis.asyncio import Redis

from app.python.auth.base import AuthContext, AuthResult, _invalidate_on_from_config


class BearerStaticAuth:
    """Static bearer token read once from an environment variable."""

    name = "bearer_static"

    def __init__(
        self,
        token: str,
        header: str,
        prefix: str,
        invalidate_on: set[int],
    ) -> None:
        self._token = token
        self._header = header
        self._prefix = prefix
        self._invalidate_on = invalidate_on

    @classmethod
    def from_config(
        cls, config: dict[str, Any], *, redis: Redis, http: httpx.AsyncClient
    ) -> BearerStaticAuth:
        if "value" in config:
            token = config["value"]
        elif "env" in config:
            env_name = config["env"]
            token = os.environ.get(env_name)
            if not token:
                raise ValueError(f"BearerStaticAuth: env var {env_name!r} is not set or empty.")
        else:
            raise ValueError("BearerStaticAuth: config must have 'env' or 'value'.")
        return cls(
            token=token,
            header=config.get("header", "Authorization"),
            prefix=config.get("prefix", "Bearer "),
            invalidate_on=_invalidate_on_from_config(config),
        )

    async def apply(self, ctx: AuthContext) -> AuthResult:
        return AuthResult(headers={self._header: f"{self._prefix}{self._token}"})

    async def invalidate(self) -> None:
        pass

    def is_rejection(self, status_code: int, body: dict | None) -> bool:
        return status_code in self._invalidate_on

File: app/ts/src/auth/modules/bearer_static.ts
import type { AuthContext, AuthResult } from "../../core/types.ts";
import type { AuthModule, AuthDeps } from "../types.ts";
import { invalidateOnFromConfig } from "../types.ts";

export class BearerStaticAuth implements AuthModule {
  static readonly typeName = "bearer_static";

  private readonly token: string;
  private readonly header: string;
  private readonly prefix: string;
  private readonly invalidateOn: Set<number>;

  private constructor(
    token: string,
    header: string,
    prefix: string,
    invalidateOn: Set<number>,
  ) {
    this.token = token;
    this.header = header;
    this.prefix = prefix;
    this.invalidateOn = invalidateOn;
  }

  static fromConfig(
    config: Record<string, unknown>,
    deps: AuthDeps,
  ): BearerStaticAuth {
    let token: string;

    if (typeof config["value"] === "string" && config["value"]) {
      token = config["value"];
    } else if (typeof config["env"] === "string") {
      const envName = config["env"];
      const envVal = deps.env[envName];
      if (typeof envVal !== "string" || !envVal) {
        throw new Error(
          `BearerStaticAuth: env var '${envName}' is not set or empty.`,
        );
      }
      token = envVal;
    } else {
      throw new Error("BearerStaticAuth: config must have 'env' or 'value'.");
    }

    const header =
      typeof config["header"] === "string" ? config["header"] : "Authorization";
    const prefix =
      typeof config["prefix"] === "string" ? config["prefix"] : "Bearer ";
    const invalidateOn = invalidateOnFromConfig(config, [401]);

    return new BearerStaticAuth(token, header, prefix, invalidateOn);
  }

  async apply(_ctx: AuthContext): Promise<AuthResult> {
    return { headers: { [this.header]: `${this.prefix}${this.token}` } };
  }

  async invalidate(): Promise<void> {}

  isRejection(statusCode: number, _body: unknown): boolean {
    return this.invalidateOn.has(statusCode);
  }
}
